### DISCUSSION TITLE
Default model: how does it work

### DISCUSSION THREAD

**Discussion by GMagician** [blakeblackshear/frigate#14292] [Ask A Question]
I'm newby about models and there is an high possibility that my question sound like an "idion" question. If there is some doc I can read to learn about how all this work it will be appreciated.
I suspect some sort of default model is used by Frigate (yolonas). I see that I have some model_cache folder and in it I see a deci-fp16-yolonas_s-rk3588-v2.0.0-1rknn file.
The little I can understand is that such model comes from something else and is "converted" to rknn.
Where "base" model is stored? Who converted it to rknn? (and how). This is a sort of "small" version and I think somewhere where is a medium/large model that should have more trained images/sizes and it may/should give better detections (but with high processing load).
N.B. I see in github some sort of notebook folder is it related to my question?
Thanks

**Comment by NickM-27** [comment]
@MarcA711 is the community contributor who has implemented the RKNN detector. The RKNN detector supports yolonas files which are converted using the rknn toolkit. Marc can provide more specific details than I can, but I believe it is the same model that is shown in the notebook in the docs

**Comment by MarcA711** [comment]
Hey,
The model used for object detection is [yolo-nas](https://github.com/Deci-AI/super-gradients/tree/master).

You are right, there are also medium and large models available. Have a look [here](https://docs.frigate.video/configuration/object_detectors#choosing-a-model).

And yes, you need to convert yolonas into a rknn format using their [toolkit](https://github.com/airockchip/rknn-toolkit2). In this repo is also a directory called docs. There are PDFs that explain the conversion process and more.

I don't use the exact same model that is produced by the notebook. The model from the notebook includes a post-processing part. This post-processing directly outputs the 20 best matches for the image. However, I had issues converting this using the rknn-toolkit2. That's why I installed super-gradients on my PC and used a small Python script (similar to the one in the notebook) to export yolonas to onnx format but without the post-processing part. The post-processing is done on the CPU with these couple of lines:
https://github.com/blakeblackshear/frigate/blob/8a8a0c7decfb0aa70cdc87d8a7de3263d4bec265/frigate/detectors/detection_api.py#L27-L73

**Comment by MarcA711** [comment]
The [rknn-model-zoo](https://github.com/airockchip/rknn_model_zoo) shows some examples of how to convert models and perform inference using the NPU of your Rockchip board.

For example [here](https://github.com/airockchip/rknn_model_zoo/tree/main/examples/yolox/python) are two scripts. One converts a yolox model from onnx format to rknn. The other one performs inference on an image using yolonas.

I wrote a little Python script that helps converting models to rknn. You can adjust all settings in a yaml file and don't need to change any python code. I have a newer version of the script that is not yet uploaded. When I find some time, I will polish my code and push them to GitHub. The old version is in this repo: https://github.com/MarcA711/rknn-models

**Reply by GMagician** [reply]
Just another question, do you know where is a good start point to learn something about how teach/build a model from images?
However thanks for all these info...

**Reply by MarcA711** [reply]
Hey,
I found this tutorial really interesting. It teaches how to build your own ai model to identify objects in images (similar to yolo-nas). It uses python + tensorflow:
https://pub.towardsai.net/building-your-own-object-detector-from-scratch-with-tensorflow-bfeadfaddad8

 This is more a hands-on tutorial to get started with this topic. If you are more interested in a article explaining how everything works in theory, I can not really recommend one.
